### PR TITLE
Experimental Cmdlets to validate objects with Azure policy conditions

### LINKED ISSUE #1266: Export Azure Policy assignments and linked definitions to a file
Related to #181, we should provide a way to export Azure Policy assignments and linked definitions to a file. 

Default to subscription scope, and exclude defintions that are not assigned.

### LINKED ISSUE #1277: Cache Azure Policy Aliases
Related to #181, we should cache Azure Policy Aliases in a JSON file

The format could be something like below:

```json
"Microsoft.AlertsManagement": {
    "smartDetectorAlertRules": {
      "locations": "global",
      "aliasMappings": {
        "Microsoft.AlertsManagement/smartDetectorAlertRules/actionGroups": "properties.actionGroups",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/actionGroups.customEmailSubject": "properties.actionGroups.customEmailSubject",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/actionGroups.customWebhookPayload": "properties.actionGroups.customWebhookPayload",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/actionGroups.groupIds": "properties.actionGroups.groupIds",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/actionGroups.groupIds[*]": "properties.actionGroups.groupIds[*]",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/description": "properties.description",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/detector": "properties.detector",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/detector.description": "properties.detector.description",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/detector.id": "properties.detector.id",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/detector.imagePaths": "properties.detector.imagePaths",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/detector.imagePaths[*]": "properties.detector.imagePaths[*]",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/detector.name": "properties.detector.name",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/detector.parameters": "properties.detector.parameters",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/detector.supportedResourceTypes": "properties.detector.supportedResourceTypes",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/detector.supportedResourceTypes[*]": "properties.detector.supportedResourceTypes[*]",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/frequency": "properties.frequency",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/scope": "properties.scope",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/scope[*]": "properties.scope[*]",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/severity": "properties.severity",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/state": "properties.state",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/throttling": "properties.throttling",
        "Microsoft.AlertsManagement/smartDetectorAlertRules/throttling.duration": "properties.throttling.duration"
      },
      "apiVersions": [
        "2018-02-01-privatepreview",
        "2019-03-01",
        "2019-06-01",
        "2021-04-01"
      ]
    }
```

Which would allow us to quickly lookup property paths for a given alias.

The other option was to use `Get-AzPolicyAlias`, but this would require an API call everytime an alias is needed, which would lead to slower performance.

### LINKED ISSUE #1278: Visit Azure Policy Assignments and Emit JSON rules
Related to #181, we should add a policy visitor class to visit exported assignments and emit JSON rules. 

We can create an `Export-AzPolicyAssignmentRuleData` cmdlet.

### LINKED ISSUE #1340: Discover Policy Assignment sources
We can include a `Get-AzPolicyAssignmentDataSource` cmdlet to discover `*.assignment.json` assignment data sources from a directory. This can search from the current working directory by default, similar to `Get-AzRuleTemplateLink`.

We could do things like `Get-AzPolicyAssignmentDataSource | Export-AzPolicyAssignmentRuleData` to export all `*.assignment,.json` sources to `*.Rule.jsonc` rules.

### PR DESCRIPTION
**Pull request by ArmaanMcleod**
## PR Summary

Fixes #1266
Fixes #1277
Fixes #1278
Fixes #1340

Added the following cmdlets:

- `Export-AzPolicyAssignmentData`
- `Export-AzPolicyAssignmentRuleData`
- `Get-AzPolicyAssignmentDataSource`

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Other code changes**
  - [x] Unit tests created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

File: tests/PSRule.Rules.Azure.Tests/Cmdlet.Common.Tests.ps1 [hunk 1]
```diff
@@ -531,18 +531,227 @@ Describe 'Get-AzRuleTemplateLink' -Tag 'Cmdlet', 'Get-AzRuleTemplateLink' {
...
+    Context 'With Defaults' {
+        BeforeAll {
+            Mock -CommandName 'GetAzureContext' -ModuleName 'PSRule.Rules.Azure' -Verifiable -MockWith ${function:MockSingleSubscription};
+
+            Mock -CommandName 'Get-AzPolicyAssignment' -ModuleName 'PSRule.Rules.Azure' -Verifiable -MockWith {
+                return @(
+                    [PSCustomObject]@{
+                        Location           = 'eastus'
+                        Name               = '000000000000000000000000'
+                        ResourceId         = '/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/PolicyRG/providers/Microsoft.Authorization/policyAssign
```
Comment (BernieWhite) [inline tests/PSRule.Rules.Azure.Tests/Cmdlet.Common.Tests.ps1:546]: @ArmaanMcleod Does this look correct?

**Review by BernieWhite — APPROVED** [review]
@ArmaanMcleod Nice work!

I made a few tweaks. There is some lines which are broken across lines, not sure if this is intended, but double check that.

Otherwise should be good to merge.

File: tests/PSRule.Rules.Azure.Tests/Cmdlet.Common.Tests.ps1 (hunk 1 above)
Comment (ArmaanMcleod) [reply tests/PSRule.Rules.Azure.Tests/Cmdlet.Common.Tests.ps1]: @BernieWhite Thanks for spotting that, missed that 😄.

**Comment by ArmaanMcleod** [comment]
> @ArmaanMcleod Nice work!
> 
> I made a few tweaks. There is some lines which are broken across lines, not sure if this is intended, but double check that.
> 
> Otherwise should be good to merge.

@BernieWhite Appreciate the tweaks and feedback, thanks for that mate 🙂

**Pull request merged**